### PR TITLE
[PrefixLM] Figuring out why prefix lm is doing poorly on short context

### PR DESCRIPTION
**Pull request by thomasw21**
What deepspeed does (I think):
 - sum microbatch losses across gradient_accumulations: https://github.com/microsoft/DeepSpeed/blob/master/deepspeed/runtime/pipe/engine.py#L671-L673
 - average loss across DP: https://github.com/microsoft/DeepSpeed/blob/master/deepspeed/runtime/pipe/engine.py#L525-L527

What Megatron-DS does:
 - averages all tokens trained on (selected via loss_masking) within microbatch (normalisation factor is the number of tokens trained on, `loss_mask.sum()`): https://github.com/bigscience-workshop/Megatron-DeepSpeed/blob/main/megatron/model/gpt_model.py#L167-L169
 
 The issue is we have a high variance of tokens we train on (`loss_mask.sum()`) in prefixlm in micro batches compared to batches. Therefore the value changes a lot and by normalising with this value we lose significant information. In order to better understand this, let us imagine two microbatches (of single samples):
 
| Microbatch id | Prefix  |  Suffix |
| - | - | - |
| 0 | t0 t1 t2 t3 t4 t5 t6 t7 t8 | t9 |
| 1 | t0 | t1 t2 t3 t4 t5 t6 t7 t8 t9 |

If you average the loss within the suffix, and then summing them together. The weight for the token t9 of MB0 is 9 times higher than then one for MB1. This means that since weights for microbatches are the same, then the model really has to make sure to get the easy samples right (ie the ones with smallest prediction AND bigger context).
 
 In order to change that, I think making the normalisation factor in Meg-DS independent of loss_mask would greatly help, since now all tokens are "equal" in terms of weight, and microbatches weight are directly dependent on the number of tokens we train on. (this code)
 
 We could also change the sampling strategy for prefix indices. Typically, weights should be like `probability -= np.arange(sequence_length, 0, -1)` where we bias it in order to have a lot more short context than long ones. cc @TevenLeScao as it was similar to an idea he had at some point.
 
 
**THIS SHOULD NOT BE MERGED AS IT CHANGES THE MODEL AND MIGHT BREAK OTHER TRAININGS**: This is just to see how one would "fix" the strong bias to long context.

### CODE REVIEW

**Comment by thomasw21** [comment]
I've updated the code to not change the behaviour of GPT. That would allow me to merge this.

**Comment by stas00** [comment]
@thomasw21, I was trying to match your solution to your description and I failed to wrap my head around. There is probably nothing wrong with it, I just understand the issue you're trying to solve.

It looks like you're approaching it in the on average each batch will have `seqlen/2` tokens and that's the expected number of tokens. 

So is this because in the case of Meg-LM we always have the full seqlen of tokens, and so you're saying that since we have to go from 1 token to seqlen tokens, then we have `(n-1)/2` tokens on average, is that the the logic?

But then it should work for the non-prefix-lm case just as well, no? It should be easy to log for both ways and compare the numbers.

Now comes the prefix-lm which I'm yet to study. So it appears that you're saying that in the case of prefix-lm there is no guarantee that we get each seqlen filled out, is that it? In which case I don't understand how `(n-1)/2` would work here.

I think it'd help to understand the problem better with a trivial example of say just 2-3 tokens and walk through the 2 scenarios.

Thank you for bearing with me.

**Comment by thomasw21** [comment]
Hey thanks for taking a look.

So quick reminder on the flow we underwent when thinking about prefix lm:

Let's take previous example we have two micro batches with 10 tokens each

| Microbatch id | sequence|
| - | - | 
| 0 | t00 t01 t02 t03 t04 t05 t06 t07 t08 t09 |
| 1 | t10 t11 t12 t13 t14 t15 t16 t17 t18 t19 |

Basically prefix lm uniformly samples an index for each sequence between 1 and sequence_length included. Everything before is defines as the prefix, and everything after is the suffix. 

| Microbatch id | prefix | suffix |
| - | - | - |
| 0 | t00 t01 t02 t03 t04 t05 t06 t07 t08 | t09 |
| 1 | t10 | t11 t12 t13 t14 t15 t16 t17 t18 t19 |

In the prefix region we allow for bidirectional attention (think Bert architectures). So naturally what we'd expected was that since tokens in that region have access to the future, predicting next token is trival. So we decided to mask it, ie compute the loss on the suffix only.

What that means is that the normalising factor we use for each MBS becomes:

| Microbatch id | prefix | suffix | loss_mask.sum() |
| - | - | - | - |
| 0 | t00 t01 t02 t03 t04 t05 t06 t07 t08 | t09 | 1 |
| 1 | t10 | t11 t12 t13 t14 t15 t16 t17 t18 t19 | 9 |

This is an issue because that means that predicting `t09` correctly has the same weight as predicting `t11 t12 t13 t14 t15 t16 t17 t18 t19` correctly. Token prediction don't have the same weight, and it seems to create a bias towards shorter predictions. In our case shorter prediction are longer prefix, which might explain our issue on prefix lm with small prefix.

Now onto the fix I propose:

So instead of normalising with a micro batch dependent value, we need to compute one that's independent of that value (replace 1 and 9 with a constant value). If it is independent, ie the same for all micro batches within a batch, then each token prediction should have the same weight.

The choice of `n/2` is linked to that mask. Though we have high variance in terms of masking per micro batch. The expected number of tokens in the suffix is about `n/2`. (It might be actually `(n+1)/2` more accurately) So instead of normalising by the number of tokens within a micro batch, we normalise with the expected value of tokens within a micro batch (that should be micro batch independent).

I think this should apply to GPT as well, but since there's no masking `loss_mask.sum()` should be the same as the product of its dimensions (full matrix of ones). I just don't want to introduce changes that might affect other experiments.

Let me know me if something is still unclear about it. Or if you notice a flaw in the reasoning!

**Comment by stas00** [comment]
That's one of the best answers I have ever read, Thomas. Thank you!

Let's do some math.

Let's calculate the normalization factor and the total loss for each microbatch and then for the global batch size

Current:

| MBS | Suffix   Loss           | Factor | T. Loss |
|   - | -                       |      - |       - |
|   0 | 10                      |      1 |      10 |
|   1 | 11 9 10                 |      3 |      10 |
|   2 | 13 8 9 11 9 10          |      6 |      10 |
|   3 | 10 10 10 13 8 9 11 9 10 |      9 |      10 |
|     |                         |        |         |

Total GBS loss: `10*4/4 = 10`

Proposed:

| MBS | Suffix   Loss           | Factor | T. Loss |
|   - | -                       |      - |       - |
|   0 | 10                      |    4.5 |     2.2 |
|   1 | 11 9 10                 |    4.5 |     6.6 |
|   2 | 13 8 9 11 9 10          |    4.5 |    13.3 |
|   3 | 10 10 10 13 8 9 11 9 10 |    4.5 |      20 |

Total GBS loss: `(2.2+6.6+13.3+20)/4 = 10.525`

It definitely checks out. 

This just requires a relatively large batch size, as for a tiny BS now the short suffixes will be greatly disfavoured.

**Comment by stas00** [comment]
p.s. our CI is currently handicapped to only runs on non-forked PR automatically, so I started the test run manually via actions:
https://github.com/bigscience-workshop/Megatron-DeepSpeed/actions/runs/1426600542

**Review by stas00 — APPROVED** [review]

**Comment by stas00** [comment]
I wonder whether `GBS/MBS` should be a factor as well (i.e. how many microbatches are involved). 

So if `GBS/MBS = 1` then factor is number of tokens
if `GBS/MBS > 10` it's n/2

**Comment by thomasw21** [comment]
Right so the ideal thing would be to average the global batch size with the amount of tokens you've backpropagated on.

Let's say you have a global batch size like the following:

| MB | Suffix   Loss           |  token count | loss.sum |
|   - | -                       |  -  | - |
|   0 | 5                      | 1 | 5 |
|   1 | 6 6 6                 | 3 | 18 |
|   2 | 7 7 7 7 7 7          | 6 | 42 |
|   3 | 10 10 10 10 10 10 10 10 10 | 9 | 90 |    

Current loss: `( 5 + 18/3 + 42/6 + 90/9 ) /  4 = 7`
Proposed loss: `( 5 + 18 + 42 + 90 ) / (9/2) / 4 = 8.61` (imagine 9 is the maximum number of tokens)
Ideal loss: `(5 + 18 + 42 + 90) / (1 + 3 + 6 + 10) = 8.6`

However the problem is we can't seem to compute the ideal loss. Why? the normalization factor `(1 +3 + 6 + 10)` requires all micro batch sizes to know how the entire global batch behaves, typically how many tokens are supposed to be trained on. With DP and GAS we don't seem to be able to (unless we accumulate the loss somewhere and then run one big backward for each global size). Anyway I tried thinking of passing weights on top of loss, but that's a breaking change for DS and I'm not sure this would be wanted.

**Comment by stas00** [comment]
So the ideal loss then would be if you don't normalize the loss at all until you got the total loss.

So the correct solution would be:
1. return un-normalized mbs loss
2. return n_tokens with the loss
3. calculate total loss with `sum(#1)/sum(#2)`

The problem is that as you're saying we don't currently have a single point where we calculate the global batch size loss and still backpropagate from there. It could be done but it'd defeat the purpose of the pipeline.

So your workaround seems to be as good as it gets solution.
